### PR TITLE
Remove var and support more general indexing

### PR DESCRIPTION
**Pull request by xukai92**
This PR removes the `Var` type (#96) and resolves the naming issue of `x[i,j,k]` and `x[i,j][k]` (#77).

Regarding the second improvement, it should allow the use any kind of nested indexing. Something like below is also supported.

```julia
using Distributions, Turing
@model test() = begin
    a = Array{Array{Float64, 2}}(2)
    for i=1:2 a[i] = Array{Float64, 2}(2, 2) end  # Init
    for i=1:2, j=1:2, k=2:3
        a[i][j, k-1] ~ Normal(0, 1)
    end
    return(a)
end
chain = @sample(test(), HMC(100, 0.2, 3));
```

### CODE REVIEW

**Comment by yebai** [comment]
Looks very good. I suggest that we rename `parse_indexing` to `VarName` and move its definition to `varinfo.jl`.

File: src/Turing.jl [hunk 1]
```diff
@@ -16,7 +16,7 @@ using Turing.Traces
 #################
 
 # Turing essentials - modelling macros and inference algorithms
-export @model, @sample, @predict, @~, @isdefined, InferenceAlgorithm, HMC, IS, SMC, PG, Gibbs, sample, Chain, Sample, Sampler, ImportanceSampler, HMCSampler, VarInfo, @predictall
+export @model, @sample, @predict, parse_indexing, @~, @isdefined, InferenceAlgorithm, HMC, IS, SMC, PG, Gibbs, sample, Chain, Sample, Sampler, ImportanceSampler, HMCSampler, VarInfo, @predictall
```
Comment (yebai) [inline src/Turing.jl]: remove obsolete export: @isdefined

File: src/core/compiler.jl [hunk 2]
```diff
@@ -2,67 +2,70 @@
...
+    else
+      # Evaluting the concrete value of the indexing variable
+      push!(uid_ex.args, quote push!(uid_list, string($evaling)) end)
+    end
+  end
+  push!(uid_ex.args, quote uid = reduce(*, uid_list) end)
+  uid_ex
+end
+
 function gen_assume_ex(left, right)
```
Comment (yebai) [inline src/core/compiler.jl]: Maybe directly place this code into macro `~`, since it is much simpler now?

Comment (xukai92) [inline src/core/compiler.jl]: Yes I agree to put them directly into `~`.

File: src/core/compiler.jl [hunk 3]
```diff
@@ -72,6 +75,8 @@ end
 
 macro ~(left, right)
```
Comment (yebai) [inline src/core/compiler.jl]: Can we store all variables in `~` and completely remove `predict` in the future?

**Comment by xukai92** [comment]
Can we merge this PR first?

**Pull request merged**